### DISCUSSION TITLE
configuring private maven / leiningen repositories

### DISCUSSION THREAD

**Discussion by thumbnail** [renovatebot/renovate#17589] [Request Help] [answered]
Hi 👋 ,

I'm setting up Renovate for a leiningen project with two private github packages registries. 
 - maven (maven.pkg.github.com)
 - docker (ghcr.io)

The private docker registry is working as intended, but i'm having problems with the maven / leiningen setup for private registries.

This is my config:

```json
{
  "extends": [
    "config:base",
    "schedule:monthly",
    "group:recommended",
    "group:allNonMajor"
  ],
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "hostRules": [
    {
      "matchHost": "ghcr.io",
      "authType": "Basic",
      "username": "github",
      "encrypted": {
        "password": "<snip>"
      }
    },
    {
      "hostType": "leiningen",
      "matchHost": "maven.pkg.github.com",
      "encrypted": {
        "token": "<snip>"
      }
    }
  ]
}
```

a redacted `project.clj`;
```clj
(defproject com.nedap.repro/project "0.0.1"
  :dependencies [[com.nedap.repro/example "1.0.0"]]

  :repositories {"central"        {:url      "https://repo1.maven.org/maven2"}
                 "clojars"        {:url      "https://clojars.org/repo"}
                 "github"         {:url      "https://maven.pkg.github.com/nedap/*"
                                   :username "github"
                                   :password :env/github_token}
                 "public-github"  {:url "git://github.com"}})
```
(the config for `:github` is working fine locally using `lein deps` btw)

The private maven repository is reporting that it isn't authenticated:
```json
DEBUG: Dependency lookup unauthorized. Please add authentication with a hostRule
{
  "failedUrl": "https://maven.pkg.github.com/nedap/*/com/nedap/repro/example/maven-metadata.xml"
}
```

Interestingly, if I curl that url with the (unencrypted) token in an `Authorization`-header it works as intended:
```bash
curl -H "Authorization: Bearer <snip>" "https://maven.pkg.github.com/nedap/*/com/nedap/repro/example/maven-metadata.xml" -v -L
# results in 200-ok with metadata
```

I also tried the following things so far:
<ol>
 <li>different matchHost options
<ul>
    <li><code>maven.pkg.github.com</code></li>
    <li><code>https://maven.pkg.github.com</code></li>
    <li><code>https://maven.pkg.github.com/nedap/*</code></li>
    <li><code>https://maven.pkg.github.com/nedap/*/</code></li>
</ul>
</li>
 <li>different hostType 
<ul>
    <li><code>leiningen</code></li>
    <li><code>maven</code></li>
</ul></li>
 <li>setup packageRules with explicit registry url:
<ul><li><details><pre>{
  "managers": [
    "maven"
  ],
  "matchPackagePrefixes": [
    "com.nedap"
  ],
  "registryUrls": [
    "https://maven.pkg.github.com/nedap/*/"
  ]
}
</pre></details></li></ul></li>
 <li>different authentication
<ul>
<li><code>Bearer</code> with a token</li>
<li><code>Basic</code> with a password</li>
<li>both options were tried with and without encryption</li>
</ul></li>
</ol>

I've also tried (most?) permutations of the options above, but no luck so far.

[this comment](https://github.com/renovatebot/renovate/issues/2158#issuecomment-399734000) mentions support for private registries, but no promises (nor in the linked PR). So i'm not sure if I should even expect this to work 🤔.

My question being: How do I configure private maven repositories for leiningen / Clojure ? Thanks!

**Comment by zharinov** [comment, marked as answer]
Seems like it should be `clojure`, not `leiningen`. Or maybe can you try to completely eliminate `hostType` if the `clojure` won't work?

**Reply by thumbnail** [reply]
🤯  it works!


Thanks, and thanks for renovate! It's working out pretty good for us so far ❤️.

**Reply by zharinov** [reply]
BTW, I like your [formatting-stack](https://github.com/nedap/formatting-stack) project ;)
